File: shared/domain/models/access.go
package models

// AccessMode は訪問記録画面における編集可否を表す。
// 仕様 docs/wants/05_チェックアウト.md「アクセスモード（読み取り／編集の二層モデル）」
type AccessMode string

const (
	AccessModeEditable AccessMode = "editable"
	AccessModeReadOnly AccessMode = "read_only"
)

// Combine は二つのアクセスモードを合成（親子関係: より制限的な側が勝つ）して返す。
// 仕様の親子関係表に従う：いずれか一方が read_only なら結果は read_only。
// 区域レベル ∩ 場所レベルの「有効モード」を求めるときに使う。
func (a AccessMode) Combine(other AccessMode) AccessMode {
	if a == AccessModeReadOnly || other == AccessModeReadOnly {
		return AccessModeReadOnly
	}
	return AccessModeEditable
}

File: pwa/src/domain/models/access.ts
// アクセスモード（読み取り／編集の二層モデル）。
// 仕様: docs/wants/05_チェックアウト.md「アクセスモード（読み取り／編集の二層モデル）」
// 参照実装: shared/domain/models/access.go

/** 訪問記録画面における編集可否。 */
export type AccessMode = "editable" | "read_only";

/**
 * 二つのアクセスモードを合成（親子関係: より制限的な側が勝つ）して返す。
 * いずれか一方が read_only なら結果は read_only。
 * 区域レベル ∩ 場所レベルの「有効モード」を求めるときに使う。
 */
export function combineAccessModes(a: AccessMode, b: AccessMode): AccessMode {
  if (a === "read_only" || b === "read_only") {
    return "read_only";
  }
  return "editable";
}
